File: Count the Number of Good Subarrays/Count the Number of Good Subarrays.py
# Solution for Count the Number of Good Subarrays in PY

class Solution:
    def countGood(self, nums: List[int], k: int) -> int:
        i, j, res, pairs = 0, 0, 0, 0
        mp = {}
        n = len(nums)
        while j < n:
            pairs += mp.get(nums[j], 0)
            mp[nums[j]] = mp.get(nums[j], 0) + 1
            while pairs >= k:
                res += n - j
                mp[nums[i]] -= 1
                pairs -= mp[nums[i]]
                i += 1
            j += 1
        return res

File: Count the Number of Good Subarrays/Count the Number of Good Subarrays.cpp
// Solution for Count the Number of Good Subarrays in CPP

class Solution {
public:
    long long countGood(vector<int>& nums, int k) {
        int n = nums.size(), i = 0, j = 0;
        long long res = 0, pairs = 0;
        unordered_map<int, int> mp;
        while (j < n) {
            pairs += mp[nums[j]];
            mp[nums[j]]++;
            while (pairs >= k) {
                res += n - j;
                mp[nums[i]]--;
                pairs -= mp[nums[i]];
                i++;
            }
            j++;
        }
        return res;
    }
};
